### PR TITLE
contrib: add direct tests for agenix 

### PR DESCRIPTION
**Pull request by ryantm**
These tests are MUCH faster than the NixOS tests.

### CODE REVIEW

**Review by n8henrie — COMMENTED** [review]
Looking through, it looks like you've migrated the tests to an installcheck on the derivation, which means they would run when the CLI was instantiated (?) / built / better word. Would we then remove the `nix flake check` process and just install the binary in CI -- if it installs, that means the tests passed?

EDIT: Premature submission. More coming.

File: pkgs/agenix.nix [hunk 1]
```diff
@@ -9,29 +9,50 @@
...
+    doInstallCheck = true;
+    installCheckInputs = [shellcheck];
+    postInstallCheck = ''
+      shellcheck ${bin}
+      ${bin} -h | grep ${version}
 
-  installPhase = ''
-    install -D $src ${placeholder "out"}/bin/agenix
-  '';
+      mkdir -p /tmp/home/.ssh
```
Comment (n8henrie) [inline pkgs/agenix.nix:34]: I assume this uses the host environment's actual `/tmp/home` (not something in the nix store)? Does this mean our tests would now have potential collisions with the host system?

File: pkgs/agenix.nix [hunk 2]
```diff
@@ -9,29 +9,50 @@
...
+      (
+        umask u=rw,g=,o=
+        cp ${../example_keys/user1} $HOME/.ssh/id_ed25519
+        chown $UID $HOME/.ssh/id_ed25519
+      )
 
-  meta.description = "age-encrypted secrets for NixOS";
-}
+      cd /tmp/home/secrets
+      ${bin} -d secret1.age | grep "hello"
```
Comment (n8henrie) [inline pkgs/agenix.nix:50]: I realize this is just a POC, but just to put it out there -- especially for something like the `-d` flag, I think we should move to something more precise than just `foo in bar` (in the old style) or `| grep foo`. For example, want to makes sure there is no leading or trailing whitespace added by an `echo` without `-n` (as opposed to `printf`) somewhere.

If sticking with bash, maybe we should write a few utility functions (that themselves should probably be tested) for this purpose, even if they're just giving a more friendly name to what is in effect `test "$(cmd)" = "foo"`.  Or maybe not, just moving to `test ... = ""` would probably be fine.

Personally, I find these kinds of tests much easier to write in Python with its tremendous stdlib. For example the tests I added re: rekeying -- awk would probably be a fine tool, but -- having written a modest amount of both at the hobbyist level -- I sure find it easier to have e.g. `first_word = output.strip().split()[0]` in my toolbox.

**Review by n8henrie — COMMENTED** [review]
Compared to the existing approach, I think this is fine. Speeding things up test cycles is nice for local development, though I don't much care if things take a minute in CI. Potential downsides include that running in a VM ensures a reproducible environment, both locally and in CI, whereas it looks like these may (?) be impure -- please help me learn if I'm wrong here.

Compared to https://github.com/ryantm/agenix/pull/159, there are a few things I like about the python module approach:

- importing from a python module lets me use a whole host of tools that I'm accustomed to when writing tests -- (limited) type checking, syntax highlighting, autocomplete
- while it will still depend on mutable state, using a python module should make is easier to break out individual tests and make them more functional, if desired -- for example isolating a failure to a specific test in a live environment would be just `tester = AgenixTester(vars); tester.run_specific_test()`
- I presume that if we ever wanted to split out tests into separate checks `checks.${system}.rekeying_words`, the python module would make this easier. Might not be much desired to ever do this though!
- I believe we could add some doctests to the tests in the python module, making a simple way to "test the test", if desired.

All that said -- the python module approach does add some complexity, take longer, and many of the above points are admittedly premature optimization! It looks like either approach should make it easier and more reliable to test the CLI as `$USER`, which I think is the first priority.

If the python module approach feels too far from the beaten path, too complex / brittle, or it's just not your preference, I won't take offense!

**Comment by n8henrie** [comment]
Also, I meant to ask -- other than speed and more directly testing user installs, are there other benefits that you envision from testing this way that I might have missed?

**Comment by ryantm** [comment]
Mainly I'm concerned about the speed of the NixOS tests.

This typically runs in the nix build sandbox, so it doesn't end up messing with the user's environment.

Another totally different path we could go is to merge our efforts with https://github.com/yaxitech/ragenix and stop coding bash, and use Rust's testing tools.

**Comment by n8henrie** [comment]
> Another totally different path we could go is to merge our efforts with https://github.com/yaxitech/ragenix and stop coding bash, and use Rust's testing tools.

I'm a big rust fanboy (though not terribly good at it yet, it's probably what I'm most focusing on learning at this point). 

What about https://github.com/ryantm/agenix/issues/23 ?

**Comment by ryantm** [comment]
My concern with #23 is that I don't want to force everyone to change their secrets.nix file to a TOML file.

File: pkgs/agenix.nix (hunk 1 above)
Comment (ryantm) [reply pkgs/agenix.nix]: I believe it uses a build sandbox, so it shouldn't collide with the host system.

**Pull request merged**